### PR TITLE
Pancakeswap migration

### PR DESCRIPTION
**Pull request by couralex6**
```
models/pancakeswap/bnb/pancakeswap_bnb_trades.sql
models/pancakeswap/ethereum/pancakeswap_ethereum_trades.sql
models/pancakeswap/pancakeswap_trades.sql
models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql
models/pancakeswap/bnb/pancakeswap_v2_bnb_mmpool_trades.sql
models/pancakeswap/bnb/pancakeswap_v2_bnb_stableswap_trades.sql
models/pancakeswap/ethereum/pancakeswap_v2_ethereum_amm_trades.sql
models/pancakeswap/ethereum/pancakeswap_v2_ethereum_mmpool_trades.sql
models/pancakeswap/bnb/pancakeswap_v3_bnb_amm_trades.sql
models/pancakeswap/ethereum/pancakeswap_v3_ethereum_amm_trades.sql
```

### CODE REVIEW

**Comment by couralex6** [comment]
This doesn't include the changes to `trace_address` and `maker`. Will fix tomorrow

File: models/pancakeswap/bnb/pancakeswap_v3_bnb_amm_trades.sql [hunk 1]
```diff
@@ -52,22 +52,22 @@ SELECT
         when lower(erc20a.symbol) > lower(erc20b.symbol) then concat(erc20b.symbol, '-', erc20a.symbol)
         else concat(erc20a.symbol, '-', erc20b.symbol)
     end as token_pair
-    ,dexs.token_bought_amount_raw / power(10, erc20a.decimals) AS token_bought_amount
-    ,dexs.token_sold_amount_raw / power(10, erc20b.decimals) AS token_sold_amount
-    ,CAST(dexs.token_bought_amount_raw AS DECIMAL(38,0)) AS token_bought_amount_raw
-    ,CAST(dexs.token_sold_amount_raw AS DECIMAL(38,0)) AS token_sold_amount_raw
+    ,CAST(dexs.token_bought_amount_raw AS UINT256) / power(10, erc20a.decimals) AS token_bought_amount
```
Comment (jeff-dude) [inline models/pancakeswap/bnb/pancakeswap_v3_bnb_amm_trades.sql:55]: any reason we explicitly cast to `unit256` in all these lines in this model, while the `v2 amm` script above doesn't? is it due to source decoded tables having differing datatypes between v2/v3?

File: models/pancakeswap/ethereum/pancakeswap_v3_ethereum_amm_trades.sql [hunk 2]
```diff
@@ -52,22 +52,22 @@ SELECT
         when lower(erc20a.symbol) > lower(erc20b.symbol) then concat(erc20b.symbol, '-', erc20a.symbol)
         else concat(erc20a.symbol, '-', erc20b.symbol)
     end as token_pair
-    ,dexs.token_bought_amount_raw / power(10, erc20a.decimals) AS token_bought_amount
-    ,dexs.token_sold_amount_raw / power(10, erc20b.decimals) AS token_sold_amount
-    ,CAST(dexs.token_bought_amount_raw AS DECIMAL(38,0)) AS token_bought_amount_raw
-    ,CAST(dexs.token_sold_amount_raw AS DECIMAL(38,0)) AS token_sold_amount_raw
+    ,CAST(dexs.token_bought_amount_raw as UINT256) / power(10, erc20a.decimals) AS token_bought_amount
```
Comment (jeff-dude) [inline models/pancakeswap/ethereum/pancakeswap_v3_ethereum_amm_trades.sql:55]: same idea on both bnb & eth it seems

**Review by jeff-dude — COMMENTED** [review]
few comments below, but otherwise this looks good to me 🙏

File: models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql [hunk 3]
```diff
@@ -21,22 +21,22 @@ WITH dexs AS
...
         t.evt_block_time                                                                AS block_time,
         t.to                                                                            AS taker,
-        ''                                                                              AS maker,
-        CASE WHEN amount0Out = '0' THEN amount1Out ELSE amount0Out END                  AS token_bought_amount_raw,
-        CASE WHEN amount0In = '0' OR amount1Out = '0' THEN amount1In ELSE amount0In END AS token_sold_amount_raw,
-        cast(NULL as double)                                                            AS amount_usd,
-        CASE WHEN amount0Out = '0' THEN f.token1 ELSE f.token0 END                      AS token_bought_address,
-        CASE WHEN amount0In = '0' OR amount1Out = '0' THEN f.token1 ELSE f.token0 END   AS token_sold_address,
+        CAST(NULL AS VARBINARY)                                                         AS maker,
+        CASE WHEN amount1Out = UINT256 '0' THEN amount1Out ELSE amount0Out END          AS token_bought_amount_raw,
```
Comment (jkylling) [inline models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql:25]: Is `amount10Out` correct here? The current version uses `amount0Out`. Same on line 28

Comment (jkylling) [reply models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql:25]: Same in other files

Comment (hsrvc) [reply models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql:25]: Should be `amount0Out` as `pancakeswap_v2_bnb.PancakePair_evt_Swap` does not have `amount10Out`

**Review by alokeutpal — APPROVED** [review]

File: models/pancakeswap/bnb/pancakeswap_v3_bnb_amm_trades.sql (hunk 1 above)
Comment (hsrvc) [reply models/pancakeswap/bnb/pancakeswap_v3_bnb_amm_trades.sql:55]: Looks like `amount` columns in v2 swap event tables are already `unit256 ` while in v3 `amount` columns are `int256`
Screenshot 2023-07-11 at 11 23 05 AM
Screenshot 2023-07-11 at 11 22 43 AM
Screenshot 2023-07-11 at 11 20 41 AM
Screenshot 2023-07-11 at 11 20 17 AM

**Comment by chef-seaweed** [comment]
Hey @couralex6 @jeff-dude, just want to double check that we are going to use `uint256` data type for `token_bought_amount_raw ` and `token_sold_amount_raw` in all tables in `dex.trades` right? thanks 🙏

**Comment by jeff-dude** [comment]
> Hi @couralex6 @jeff-dude, just want to double check that we are going to use `uint256` data type for `token_bought_amount_raw` and `token_sold_amount_raw` in all tables in `dex.trades` right? thanks 🙏

this was something we were just discussing internally too. the takeaway seems to be `uint256` for sure. i'm curious though, if you don't mind, can you explain the why from your perspective? or how that works from the smart contract perspective in terms of older versions/newer versions changing types between `int256` and `uint256`?

**Comment by chef-seaweed** [comment]
> this was something we were just discussing internally too. the takeaway seems to be `uint256` for sure.
Got it, thanks 🙏

> i'm curious though, if you don't mind, can you explain the why from your perspective? or how that works from the smart contract perspective in terms of older versions/newer versions changing types between `int256` and `uint256`?
I think in this regard we just follow what the original Uniswap v2/v3 contracts do. But I think we can refer to the answer I received in Uniswap discord as shown in the attached pic.

Screenshot 2023-07-12 at 10 35 23 AM

File: models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql [hunk 4]
```diff
@@ -1,11 +1,11 @@
-{{ config(
+{{ config(tags=['dunesql'],
     schema = 'pancakeswap_v2_bnb',
     alias = alias('amm_trades'),
     partition_by = ['block_date'],
```
Comment (jeff-dude) [inline models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql:4]: until we finalize strategy here: https://dune.height.app/T-31748

we've been changing partitions on migrated spells to `block_month` to grow the partition sizes.

can you add a new column in the final select, right below `block_date`, where we trunc to month and output `block_month`, then apply to partition?

File: models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql [hunk 5]
```diff
@@ -21,22 +21,21 @@ WITH dexs AS
...
-        CASE WHEN amount0Out = '0' THEN amount1Out ELSE amount0Out END                  AS token_bought_amount_raw,
-        CASE WHEN amount0In = '0' OR amount1Out = '0' THEN amount1In ELSE amount0In END AS token_sold_amount_raw,
-        cast(NULL as double)                                                            AS amount_usd,
-        CASE WHEN amount0Out = '0' THEN f.token1 ELSE f.token0 END                      AS token_bought_address,
-        CASE WHEN amount0In = '0' OR amount1Out = '0' THEN f.token1 ELSE f.token0 END   AS token_sold_address,
+        CAST(NULL AS VARBINARY)                                                         AS maker,
+        CASE WHEN amount0Out = UINT256 '0' THEN amount1Out ELSE amount0Out END          AS token_bought_amount_raw,
+        CASE WHEN amount0In = UINT256 '0' OR amount1Out = UINT256 '0' THEN amount1In ELSE amount0In END AS token_sold_amount_raw,
+        NULL                                                            AS amount_usd,
+        CASE WHEN amount1Out = UINT256 '0' THEN f.token1 ELSE f.token0 END              AS token_bought_address,
```
Comment (jeff-dude) [inline models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql:28]: typo to change this from `amount0Out` to `amount1Out`?
looks like line 25 above is using `amount0Out` too

File: models/yearn/optimism/yearn_optimism_vaults.sql
Comment (jeff-dude) [inline models/yearn/optimism/yearn_optimism_vaults.sql:1]: is this file accidental? potentially remove?

File: models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql (hunk 5 above)
Comment (jeff-dude) [reply models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql:28]: also looks like ethereum version keep `amount0`

File: models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql (hunk 4 above)
Comment (jeff-dude) [reply models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql:4]: this will apply to all materialized spells in here that are partitioned

**Review by jeff-dude — CHANGES_REQUESTED** [review]

**Comment by jeff-dude** [comment]
@Hosuke same idea here as sushiswap
https://github.com/duneanalytics/spellbook/pull/3686#issuecomment-1662940493

File: models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql (hunk 5 above)
Comment (couralex6) [reply models/pancakeswap/bnb/pancakeswap_v2_bnb_amm_trades.sql:28]: Oh good catch. I had to fix this for Uniswap too. Must be a search and replace error

**Comment by Hosuke** [comment]
> @Hosuke same idea here as sushiswap [#3686 (comment)](https://github.com/duneanalytics/spellbook/pull/3686#issuecomment-1662940493)

Added `block_month` as a new column and recovered files.

File: models/yearn/optimism/yearn_optimism_vaults.sql
Comment (couralex6) [reply models/yearn/optimism/yearn_optimism_vaults.sql:1]: This is in the pancakeswap lineage

Comment (jeff-dude) [reply models/yearn/optimism/yearn_optimism_vaults.sql:1]: got it. if not in PR, it'll grab it from prod for CI tests, so no big deal if missing here with no change needed

**Review by jeff-dude — APPROVED** [review]
lfg

**Pull request merged**